### PR TITLE
[docs] redesign: draft of Layout Header with some improvements

### PR DESCRIPTION
**Pull request by Simek**
# Why

This PR ports the draft of the `Header` component for the new `Layout` from the Cedric fork and adds few fixes and tweaks in the code.

# How

The component should work more-less as expected, but the Layout still doesn't not function 100% correct on the mobile devices. It is a step forward, but still a solution which is not production ready (currently is only used on the new Home page). 

There are some (probably not all) issue on mobile I can see now:
* there is still something weird with overflow and grid rows height on mobile, I have spotted that two scrollbars are overlapped, one from the nested ScrollLayout, and second for the whole page (which should not be there), if I disable that manually, then con tent row is clipped at the bottom, 🤷‍♂️ 
* It's very tricky to fill the whole screen with search results on mobile (as in design), mainly due to sliding up keyboard, which breaks all the viewport based calculations.

So, feel free to take over or modify code of this PR, my experience with CSS grids isn't that broad, so sorry if I made any incorrect changes, but I have tried to do my best in there. 😅 

# Test Plan

Running the docs website on localhost and using the new Home page as a test bench.

# Preview

## Desktop

Screenshot 2022-03-21 at 16 31 32

## Mobile

https://user-images.githubusercontent.com/719641/159296343-19e62e59-06f4-4498-8e16-3db85f60728c.MOV

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by Simek** [comment]
The latest updates switches the Layout implementation from CSS Grids to Flexbox, which seems to be way less problematic approach. Additionally I have updated the HTML semantics and I have tried to simplify the code as much as possible and remove all of the redundant styling.

After the update I was only able to find small issue on iOS on Chrome, where site scrollbar has a bit offset to the correct position, this do not happen in Safari nor Firefox Focus. On desktop the site seems to be fine across all browsers I use on macOS and Win.

So beside this small UI glitch, the PR seems to be ready for the review & merge. The next step will be porting new 404 and 500 pages, which after the redesign requires the header.

# Preview

File: docs/ui/components/Header/Header.tsx [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        <nav css={[containerStyle, showOnMobileStyle]}>
+          <div css={[columnStyle, leftColumnStyle]}>
+            <BOLD>Theme</BOLD>
+          </div>
+          <div css={[columnStyle, rightColumnStyle]}>
+            <ThemeSelector />
+          </div>
+        </nav>
+      ) : null}
+      {isMobileSearchVisible ? (
```
Comment (byCedric) [inline docs/ui/components/Header/Header.tsx:57]: Since this is a boolean, you can also use something like:

```
{isMobileSearchVisible && (
  ...
)}
```

But that's like a personal preference, not mandatory 😄

File: docs/ui/components/Header/Search.tsx [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+import { paragraph } from '~/components/base/typography';
+import { LATEST_VERSION } from '~/constants/versions';
+import { SlashShortcutIcon } from '~/ui/foundations/icons';
+
+type Props = {
+  version: string;
+  mobile?: boolean;
+};
+
+// TODO(simek): ensure that processUrl is still needed
```
Comment (byCedric) [inline docs/ui/components/Header/Search.tsx:16]: Good to check indeed, I think we don't need it anymore. But maybe NextJS's router does some weird stuff with absolute URLs, as in, it probably refreshes the whole page instead of client-side redirect.

File: docs/ui/components/Layout/Layout.tsx [hunk 3]
```diff
@@ -1,67 +1,80 @@
...
   },
 });
 
-const contentStyle = css({ gridArea: 'content' });
 const innerContentStyle = css({
   margin: '0 auto',
   maxWidth: breakpoints.large,
-  height: '100%',
-  overflowY: 'visible',
+  padding: `${spacing['8']}px ${spacing['4']}px`,
```
Comment (byCedric) [inline docs/ui/components/Layout/Layout.tsx:73]: Oh man, I'd really wish we had this on web too 😂 

```js
{
  paddingVertical: spacing[8],
  paddingHorizontal: spacing[4],
}
```

File: docs/ui/components/Layout/LayoutScroll.tsx [hunk 4]
```diff
@@ -4,37 +4,49 @@ import React, { forwardRef, HTMLAttributes, PropsWithChildren } from 'react';
 
 type LayoutScrollProps = PropsWithChildren<
   HTMLAttributes<HTMLDivElement> & {
-    /** If the scroll container should smoothly scroll when scrolled programatically */
+    // If the scroll container should smoothly scroll when scrolled programatically
```
Comment (byCedric) [inline docs/ui/components/Layout/LayoutScroll.tsx:7]: Why did you change this from jsdoc to inline comment? I think it's part of the props, so jsdoc would be better here (bc vscode will just add this description to hovering over that prop when using the component)

**Review by byCedric — APPROVED** [review]
1 question related to jsdoc to inline comment, but overall looks good to me. It's unfortunate that the grid system is causing some issues on mobile though.

File: docs/ui/components/Header/Header.tsx (hunk 1 above)
Comment (Simek) [reply docs/ui/components/Header/Header.tsx:57]: I was using this notation for a bit, but I have decided to switch back to full ternary:
* https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx

File: docs/ui/components/Layout/LayoutScroll.tsx (hunk 4 above)
Comment (Simek) [reply docs/ui/components/Layout/LayoutScroll.tsx:7]: I thought that since this type is not exported the comments in there are just notes, not the actual docs. 😅 

I will change back the comment format.

**Review by jonsamp — APPROVED** [review]
This seems good to me, thanks for porting this over with fixes

File: docs/ui/components/Header/Search.tsx [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+const processUrl = (url: string) => {
+  // Update URLs for new doc URLs
+  const routes = url.split('/');
+  routes[routes.length - 1] = routes[routes.length - 1].replace('.html', '');
+  return routes.join('/');
+};
+
+export class Search extends React.Component<Props> {
```
[the remark this answers is not in the document]
Comment (Simek) [reply docs/ui/components/Header/Search.tsx:24]: AFAIK, we also will need to switch to new Algolia search soon, which will change the whole search UI and UX, so maybe it's worth to do the conversion and switch to new API at the same time.

File: docs/ui/components/Header/Search.tsx (hunk 2 above)
Comment (Simek) [reply docs/ui/components/Header/Search.tsx]: I have removed this function after Search conversion to Functional Component, and everything seems to work correctly.

File: docs/ui/components/Header/Header.tsx (hunk 1 above)
Comment (Simek) [reply docs/ui/components/Header/Header.tsx]: Since using `&&` do not pose any potential issues, I have decided to apply the suggestion! 🙂

File: docs/ui/components/Layout/LayoutScroll.tsx (hunk 4 above)
Comment (Simek) [reply docs/ui/components/Layout/LayoutScroll.tsx:7]: It's done.

File: docs/ui/components/Header/Logo.tsx [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+`;
+
+const logoStyle = css`
+  float: left;
+  margin: 2px 8px 0 0;
+`;
+
+const chevronStyle = css`
+  transform: rotate(-90deg);
+  color: ${theme.icon.secondary};
```
[the remark this answers is not in the document]
Comment (Simek) [reply docs/ui/components/Header/Logo.tsx]: I have confirm this locally, and as you suggested the color in there was redundant. 🙂

File: docs/ui/components/Header/Search.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (Simek) [reply docs/ui/components/Header/Search.tsx]: Since there were not that much code, I have decided to convert the Search component now anyway, so it's not a reason to block this PR.

**Comment by Simek** [comment]
All suggested changes have been applied, but since the search component has been refactored and the styles changed, it would be nice to re-review the PR.

**Review by byCedric — APPROVED** [review]
Looking good!

**Review by jonsamp — APPROVED** [review]
